### ISSUE TITLE
Set up CI

### ISSUE THREAD

**Issue by vincerubinetti** [dashnowlab/STRchive#59]
I believe this is what our GitHub Actions CI workflow should eventually look like:

```yaml
name: Update data

on:
  pull_request:
  workflow_dispatch:

jobs:
  update:
    runs-on: ubuntu-latest

    steps:
      - name: Debug dump
        uses: crazy-max/ghaction-dump-context@v2

      - name: Checkout code
        uses: actions/checkout@v4

      - name: Setup Python
        uses: actions/setup-python@v5
        with:
          python-version: "3.11"
          cache-dependency-path: "**/environment.yml"

      - name: Install Python packages
        run: |
          echo Some conda install command

      - name: SSH debug
        if: runner.debug == '1'
        uses: mxschmitt/action-tmate@v3

      - name: Update data
        run: |
          python some-script-1.py
          python some-script-2.py
          python some-script-3.py

      # optional. might be nice for the site to know when the data was last updated. 
      - name: Record data compile time
        run: sed -i "" "s/PUBLIC_DATA_DATE=*/PUBLIC_DATA_DATE=$(date -uIseconds)/" site/.env

      - name: Open pull request with updated files
        if: github.event_name == 'workflow_dispatch'
        uses: peter-evans/create-pull-request@v6
        with:
          branch: update-data
          title: Update data

      - name: Commit updated files
        if: github.event_name == 'pull_request'
        uses: stefanzweifel/git-auto-commit-action@v5
        with:
          commit_message: Update data
```

This will allow you to either manually run the workflow and have it open a PR with the updated data, or open a PR manually and have it run the data update automatically on the PR.

As for the site, I'm thinking that we should use Netlify instead of GitHub Pages for the new website. They are both free and easy to use, but Netlify also gives us live deploy previews of PRs, built in. You can of course also set up your Netlify site to use your custom domain. And we'll have it configured such that it just rebuilds and redeploys the site when there's any changes in the repo (on main or any PR branch), including the `/data` folder. As such, there's no need to trigger anything related to the site in this gh-actions workflow; it will happen automatically. Yes this means that the site will be rebuilt when ineffectual things like the readme change, but the cost is minimal; the site only takes a few seconds to build.

**Comment by vincerubinetti** [comment]
Not sure what you had planned, but it seemed to me like maybe all the data should be updated at the same time, i.e. as part of the same gh-actions workflow. Unless you want to be able to, say, update the literature independently from the other stuff. That might muddy the waters though, like the literature would be on its own version (if it even has a version) separate from the rest of the data.

> version the json data and link to it, like what I did manually here:

For reference, here's the "versioning" workflow I have for Lab Website Template: https://github.com/greenelab/lab-website-template/blob/main/.github/workflows/versioning.yaml [`ncipollo/release-action`](https://github.com/ncipollo/release-action) makes it easy to make tags and releases.

> This should also appear at the top of the table and somewhere on the locus pages along with an updated date.

If what I said above is what you decide to do, there'd be one version for all of the data, and that version could be displayed in the header perhaps, with a link to the list of releases / changelog.

Instead of the "optional record data compile time" step I put above, it'd actually be better if you made a [GitHub CFF citation file](https://github.com/greenelab/lab-website-template/blob/main/CITATION.cff) for this repo. You probably should have this anyway. But it will also allow me to conveniently get the version and date of the data to display on the website somewhere.

> Run scripts to update locus definitions (probably should generate a PR and be checked)
Update literature for existing loci (scheduled, auto PR)
Search for new locus literature (scheduled, auto PR)

I'd imagine all these scripts would run in sequence in this same workflow, and then the workflow could be triggered by `workflow_dispatch` (running it with a manual button click in the GitHub web interface), `pull_request` (when you manually open a PR for whatever reason), and `schedule` (perhaps weekly or monthly). Then it would open a PR with [`peter-evans/create-pull-request`](https://github.com/peter-evans/create-pull-request) for you to review and merge. We could have it always open a new PR, or give a specific branch name such that if one is already open (e.g. you haven't gotten around to merging last week's update PR yet), it just updates that one.

**Comment by laurelhiatt** [comment]
Hey Vince sorry I didn't see this, I thought CI was confidence interval for the plots lol, which is out of my jurisdiction. we can meet about this tomorrow or this week if you'd like

**Comment by vincerubinetti** [comment]
<img width="1256" alt="Screenshot 2024-11-26 at 1 09 09 PM" src="https://github.com/user-attachments/assets/518eaa43-ff0a-4f60-b738-ad7feebd64d5">

**Comment by hdashnow** [comment]
#93 should have all the components needed for automation. See the `STRchive/README.md` and let me know if more details are needed.
Note that the `run-manubot.py` script runs so long (12+ hours) that I've never actually run it on the full dataset, only a subset. I think we'll need to deal with this before we can use it in CI. In the meantime, you can run everything else using `snakemake --config stages="skip-refs"` in a few seconds.

**Comment by vincerubinetti** [comment]
It probably will always take at least a couple of hours if doing every citation, but i would look into running things in parallel to see if it improves times. The manubot script I gave you just runs one at a time. You can also pass multiple IDs to manubot in one command, though idk if it parallelizes that.

**Comment by hdashnow** [comment]
I thought about doing things in parallel. It could cause rate limit issues if not done carefully. I've got it down to ~2 hours run time now, so let's see how we go and decide if it's worth the effort to find more speed-ups.

**Comment by vincerubinetti** [comment]
```yaml
name: Update data

on:
  pull_request:
    branches: main
    paths:
      - "data/**"
      - "scripts/**"
  schedule:
    - cron: "0 0 1 * *"
  workflow_dispatch:

jobs:
  update:
    runs-on: ubuntu-latest

    steps:
      - name: Debug dump
        uses: crazy-max/ghaction-dump-context@v2

      - name: Checkout code
        uses: actions/checkout@v4

      - name: SSH debug
        if: runner.debug == '1'
        uses: mxschmitt/action-tmate@v3

      - name: Set up Python
        uses: actions/setup-python@v5
        with:
          python-version: "3.12"
          cache-dependency-path: "**/environment.yml"

      - name: Set up Conda
        uses: s-weigand/setup-conda@v1

      - name: Activate Conda
        run: |
          conda env create --file scripts/environment.yml
          conda init
          source ~/.zshrc
          conda activate strchive

      - name: Update data (short)
        if: ${{ github.event_name == 'pull_request' }}
        run: snakemake --config stages="skip-refs"

      - name: Update data (full)
        if: ${{ github.event_name == 'schedule' || github.event_name == 'workflow_dispatch' }}
        run: snakemake

      - name: Open pull request with updated files
        if: ${{ !(github.event_name == 'pull_request') }}
        uses: peter-evans/create-pull-request@v7
        with:
          branch: update-data
          title: Update data

      - name: Commit updated files to current pull request
        if: ${{ github.event_name == 'pull_request' }}
        uses: stefanzweifel/git-auto-commit-action@v5
        with:
          commit_message: Update data
```

**Comment by hdashnow** [comment]
I'm out of my depth on this CI stuff... This worked on my fork, but can't get it working in this repository.
https://github.com/dashnowlab/STRchive/actions/runs/12246554105

**Comment by vincerubinetti** [comment]
Here's the failing run:

https://github.com/dashnowlab/STRchive/actions/runs/12246554105/job/34162663499

And the relevant logs:

```
remote: Permission to dashnowlab/STRchive.git denied to github-actions[bot].
fatal: unable to access 'https://github.com/dashnowlab/STRchive/': The requested URL returned error: 403
```

It's failing with a [403 permission denied error code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses). So the action runner (i.e. the "github actions bot") doesn't have permissions to commit/push to the repo.

I believe you'll need to allow actions read + write permissions in the repo settings:

<img width="1210" alt="Screenshot 2024-12-10 at 12 33 52 PM" src="https://github.com/user-attachments/assets/8d4e1fa7-3f40-4a28-8afa-b3ad2908bfdf">

Hopefully if you change that and re-run the failed workflow, it should work.

**Comment by hdashnow** [comment]
Oops, closed prematurely. Still working on the R bit of this.
